daemon/mgr/container.go

```diff
@@ -1331,6 +1341,378 @@ func (mgr *ContainerManager) updateContainerResources(c *Container, resources ty
+	}
+
+	// We need to rebase the archive entries if the last element of the
+	// resolved path was a symlink that was evaluated and is now different
+	// than the requested path. For example, if the given path was "/foo/bar/",
+	// but it resolved to "/var/lib/pouch/containers/{id}/foo/baz/", we want
+	// to ensure that the archive entries start with "bar" and not "baz". This
+	// also catches the case when the root directory of the container is
+	// requested: we want the archive entries to start with "/" and not the
+	// container ID.
+	data, err := archive.TarResourceRebase(resolvedPath, filepath.Base(absPath))
+	if err != nil {
+		return nil, nil, err
+	}
+
+	content = ioutils.NewReadCloserWrapper(data, func() error {
+		err := data.Close()
+		return err
+	})
+
+	mgr.LogContainerEvent(ctx, container, "archive-path")
+
+	return content, stat, nil
+}
+
+// ResolvePath resolves the given path in the container to a resource on the
+// host. Returns a resolved path (absolute path to the resource on the host),
+// the absolute path to the resource relative to the container's baseFS, and
+// an error if the path points to outside the container's baseFS.
+func ResolvePath(container *Container, path string) (resolvedPath, absPath string, err error) {
+	absPath = archive.PreserveTrailingDotOrSeparator(filepath.Join(string(filepath.Separator), path), path)
+
+	// Split the absPath into its Directory and Base components. We will
+	// resolve the dir in the scope of the container then append the base.
+	dirPath, basePath := filepath.Split(absPath)
+
+	resolvedDirPath, err := container.GetResourcePath(dirPath)
+	if err != nil {
+		return "", "", err
+	}
+
+	// resolvedDirPath will have been cleaned (no trailing path separators) so
+	// we can manually join it with the base path element.
+	resolvedPath = resolvedDirPath + string(filepath.Separator) + basePath
+
+	return resolvedPath, absPath, nil
+}
+
+// ContainerStatPath stats the filesystem resource at the specified path in the
+// container identified by the given name.
+func (mgr *ContainerManager) ContainerStatPath(ctx context.Context, name string, path string) (stat *types.ContainerPathStat, err error) {
+	container, err := mgr.container(name)
+	if err != nil {
+		return nil, err
+	}
+
+	return mgr.containerStatPath(ctx, container, path)
+}
+
+// containerStatPath stats the filesystem resource at the specified path in this
+// container. Returns stat info about the resource.
+func (mgr *ContainerManager) containerStatPath(ctx context.Context, c *Container, path string) (stat *types.ContainerPathStat, err error) {
+	//There is a dead lock head
+	//c.Lock()
+	//defer c.Unlock()
+
+	if err := mgr.Mount(ctx, c); err != nil {
+		return nil, err
+	}
+	defer mgr.Unmount(ctx, c)
+
+	attachedVolumes := map[string]struct{}{}
+	defer func() {
+		if err == nil {
+			return
+		}
+
+		// release the container resources(network and containerio)
+		err = mgr.releaseContainerResources(c)
```

Comment:
I don't think it is good move for release. we cannot impact running container if pouch fails to do head action.